lib/index.js

```diff
@@ -0,0 +1,507 @@
+        return cb(err)
+      }
+
+      id = row.key.split('!')[1]
+      this.remove(id, cb)
+    })
+    return
+  }
+
+  this.db.del(getKey(id), cb)
+}
+
+Clocker.prototype.add = function (start, end, data, cb) {
+  var from = strftime('%F %T', getDate(start))
+
+  if (typeof data === 'function') {
+    cb = data
+    data = {}
+  }
+
+  var value = {}
+  for (var key in data) {
+    value[key] = data[key]
+  }
+  value.end = getDate(end)
+
+  var type = value.type
+  var pkey = 'time!' + from
+  var tkey = 'time-type!' + type + '!' + from
+
+  this.db.batch([
+    { type: 'put', key: pkey, value: value },
+    { type: 'put', key: tkey, value: 0 }
+  ], function (err) {
+    if (err) {
+      return cb(err)
+    }
+
+    var stamp = toStamp(pkey)
+    return cb(null, stamp)
+  })
+}
+
+Clocker.prototype.dataStream = function (filter) {
+  filter = filter || {}
+  filter.test = filter.test || function () { return true }
+
+  var s = this.db.createReadStream({
+    gt: 'time!' + (filter.gt ? strftime('%F %T', getDate(filter.gt)) : ''),
+    lt: 'time!' + (filter.lt ? strftime('%F %T', getDate(filter.lt)) : '~')
+  })
+
+  var stream = s.pipe(through2.obj(function (row, _, cb) {
+    var entry = Clocker.getDataObject(row)
+    cb(null, entry)
+  })).pipe(through2.obj(function (entry, _, cb) {
+    if (filter.test(entry)) {
+      cb(null, entry)
+    } else {
+      cb()
+    }
+  }))
+
+  return stream
+}
+
+Clocker.prototype.data = function (filter, cb) {
+  if (typeof filter === 'function') {
+    cb = filter
+    filter = {}
+  }
+
+  var data = []
+  this.dataStream(filter).on('error', function (err) {
+    return cb(err)
+  }).on('end', function () {
+    return cb(null, data)
+  }).pipe(through2.obj(function (entry, _, finish) {
+    data.push(entry)
+    finish()
+  }))
+}
+
+Clocker.prototype.aggregate = function (by, filter, cb) {
+  if (typeof filter === 'function') {
+    cb = filter
+    filter = {}
+  }
+
+  var data = {}
+  this.dataStream(filter).on('error', function (err) {
+    return cb(err)
+  }).on('end', function () {
+    return cb(null, data)
+  }).pipe(through2.obj(function (entry, _, finish) {
+    // entry might correspond to different by's
+
+    var byValues = getBys(entry, by)
+    byValues.forEach(function (byValue) {
+      if (!data.hasOwnProperty(byValue[0])) {
+        data[byValue[0]] = 0
+      }
+      data[byValue[0]] += byValue[1]
+    })
+
+    finish()
+  }))
+}
+
+Clocker.prototype._getLastRow = function (cb) {
+  var found = false
+  this.db.createReadStream({
+    gt: 'time!',
+    lt: 'time!~',
+    limit: 1,
+    reverse: true
+  }).once('data', function (data) {
+    found = true
+    return cb(null, data)
+  }).on('error', function (err) {
+    return cb(err)
+  }).on('end', function () {
+    if (found === false) {
+      return cb(new Error('Empty database'))
+    }
+  })
+}
+
+function getBys (data, _by) {
+  var day = {
+    get: (e) => {
+      var p = e.toLocaleDateString().split('/')
```

Comment:
This does not work on my machine because of the time local. `e` is something like: `2018-01-01T07:00:00.000Z` and therefore the split using `/` does not work. And the order of `p[2]`, `p[0]` and `p[1]` does not work later on.

But the good thing: The unit tests found it! 🎉 
```
# aggregate
# by day
# simple not overlapping days
e 2018-01-01T07:00:00.000Z
p [ '2018-1-1' ]
/Users/beni/Documents/git/BeniRupp/clocker/lib/index.js:451
      return p[2] + '-' + p[0].padStart(2, '0') + '-' + p[1].padStart(2, '0')
```

Reply:
Thank you for spotting this. Fixed it by using `Date.toISOString()` (plus timezone handling) instead.